File: 백준/Silver/10971. 외판원 순회 2/외판원 순회 2.py
import sys
input = sys.stdin.readline

ans = sys.maxsize


def dfs(start, next, cost, visited):
    global ans
    if len(visited) == N:
        if board[next][start]:
            ans = min(ans, cost+board[next][start])
        return

    for i in range(N):
        if board[next][i] and i not in visited and not ans < cost:
            visited.append(i)
            dfs(start, i, cost+board[next][i], visited)
            visited.pop()


N = int(input())
board = [list(map(int, input().split())) for _ in range(N)]

for i in range(N):
    dfs(i, i, 0, [i])

print(ans)

File: 백준/Silver/10971. 외판원 순회 2/외판원 순회 2.java

import java.util.Scanner;

public class Main {
    static int ans = Integer.MAX_VALUE;
    static int[][] board;
    static int N;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        board = new int[N][N];
        
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                board[i][j] = sc.nextInt();
            }
        }

        for (int i = 0; i < N; i++) {
        	boolean[] visited = new boolean[N];
        	visited[i] = true;
            dfs(i, i, 0, visited, 1);
        }

        System.out.println(ans);
    }

    static void dfs(int start, int next, int cost, boolean[] visited, int cnt) {
        if (cnt == N) {
            if (board[next][start] != 0) {
                ans = Math.min(ans, cost + board[next][start]);
            }
            return;
        }

        for (int i = 0; i < N; i++) {
            if (board[next][i] != 0 && !visited[i] && !(ans < cost)) {
                visited[i] = true;
                dfs(start, i, cost + board[next][i], visited, cnt+1);
                visited[i] = false;
            }
        }
    }
}
